File: backend/src/main/java/by/fpmibsu/findafriend/controller/controllers/UserController.java
package by.fpmibsu.findafriend.controller.controllers;

import by.fpmibsu.findafriend.application.HandleResult;
import by.fpmibsu.findafriend.application.controller.*;
import by.fpmibsu.findafriend.application.mediatr.Mediatr;
import by.fpmibsu.findafriend.controller.AuthUtils;
import by.fpmibsu.findafriend.controller.Logging;
import by.fpmibsu.findafriend.controller.Validation;
import by.fpmibsu.findafriend.controller.commands.users.CreateUserCommand;
import by.fpmibsu.findafriend.controller.commands.users.DeleteUserCommand;
import by.fpmibsu.findafriend.controller.commands.users.SetUserRoleCommand;
import by.fpmibsu.findafriend.controller.commands.users.UpdateUserCommand;
import by.fpmibsu.findafriend.controller.queries.users.GetUserByIdQuery;
import by.fpmibsu.findafriend.controller.queries.users.GetUsersQuery;
import by.fpmibsu.findafriend.dataaccesslayer.user.UserDao;
import by.fpmibsu.findafriend.entity.User;
import org.apache.logging.log4j.LogManager;
import org.apache.logging.log4j.Logger;

import java.util.Arrays;

@ControllerRoute(route = "/users")
public class UserController extends Controller {
    private static final Logger logger = LogManager.getLogger(UserController.class);
    private final Mediatr mediatr;

    public UserController(Mediatr mediatr) {
        this.mediatr = mediatr;
    }

    @Endpoint(path = "/getAll", method = HttpMethod.GET)
    public HandleResult getAll() {
        return ok(mediatr.send(new GetUsersQuery()));
    }

    @RequireAuthentication
    @Endpoint(path = "/getById", method = HttpMethod.GET)
    public HandleResult getById(@FromQuery(parameterName = "id") int id) {
        return ok(mediatr.send(new GetUserByIdQuery(id)));
    }

    @Endpoint(path = "/create", method = HttpMethod.POST)
    public HandleResult createUser(@FromBody CreateUserCommand request) {
        if (isAnyEmpty(request.name, request.surname, request.login, request.password, request.email, request.phoneNumber)) {
            return badRequest();
        }
        if (request.password.length() < 8) {
            return badRequest("Too short password");
        }
        var userDao = serviceProvider.getRequiredService(UserDao.class);
        if (userDao.findByLogin(request.login) != null) {
            return badRequest("Duplicate login");
        }
        return ok(mediatr.send(request));
    }

    @RequireAuthentication
    @Endpoint(path = "/update", method = HttpMethod.PUT)
    public HandleResult updateUser(@FromBody UpdateUserCommand request, @WebToken(parameterName = "id") int userId,
                                   @WebToken(parameterName = "role") String role) {
        if (userId != request.userId && !AuthUtils.allowRoles(role, User.Role.ADMINISTRATOR)) {
            Logging.warnNonAuthorizedAccess(this.request, logger);
            return notAuthorized();
        }
        return ok(mediatr.send(request));
    }

    @RequireAuthentication
    @Endpoint(path = "/delete", method = HttpMethod.DELETE)
    public HandleResult deleteUserById(@FromQuery(parameterName = "id") int id, @WebToken(parameterName = "id") int userId,
                                       @WebToken(parameterName = "role") String role) {
        if (userId != id && !AuthUtils.allowRoles(role, User.Role.ADMINISTRATOR)) {
            Logging.warnNonAuthorizedAccess(this.request, logger);
            return notAuthorized();
        }
        return ok(mediatr.send(new DeleteUserCommand(id)));
    }

    @RequireAuthentication
    @Endpoint(path = "/getSelfInfo", method = HttpMethod.GET)
    public HandleResult getSelfInfo(@WebToken(parameterName = "id") int userId) {
        return getById(userId);
    }

    @RequireAuthentication
    @Endpoint(path = "/setRole", method = HttpMethod.PUT)
    public HandleResult updateRole(@FromBody SetUserRoleCommand command,
                                   @WebToken(parameterName = "role") String role) {
        if (!Validation.in(command.newRole, User.Role.USER.toString(), User.Role.MODERATOR.toString(),
                User.Role.SHELTER_ADMINISTRATOR.toString(), User.Role.ADMINISTRATOR.toString())) {
            return badRequest();
        }
        if (!AuthUtils.allowRoles(role, User.Role.ADMINISTRATOR)
                || AuthUtils.allowRoles(command.newRole, User.Role.ADMINISTRATOR)) {
            return notAuthorized();
        }

        var user = serviceProvider.getRequiredService(UserDao.class).getEntityById(command.userId);
        if (user == null) {
            return badRequest();
        }
        if (User.Role.ADMINISTRATOR.equals(user.getRole())) {
            return notAuthorized();
        }
        return ok(mediatr.send(command));
    }

    private static boolean isAnyEmpty(String... strings) {
        return Arrays.stream(strings)
                .anyMatch(s -> s == null || "".equals(s));
    }
}

File: BackendNET/Backend/Controllers/UserController.cs
using Backend.Commands.Users.CreateUser;
using Backend.Commands.Users.DeleteUser;
using Backend.Commands.Users.SetUserRole;
using Backend.Commands.Users.UpdateUser;
using Backend.Dto;
using Backend.Queries.Users.GetAllUsers;
using Backend.Queries.Users.GetIdAndLogin;
using Backend.Queries.Users.GetUserById;
using Backend.Services;
using Backend.Utils;
using MediatR;
using Microsoft.AspNetCore.Authorization;
using Microsoft.AspNetCore.Mvc;
using static Backend.Entities.User;

namespace Backend.Controllers;

public class UserController : ApiController
{
    private readonly IMediator _mediator;
    private readonly IUserService _userService;

    public UserController(IMediator mediator, IUserService userService)
    {
        _mediator = mediator;
        _userService = userService;
    }

    [HttpGet("/users/getAll")]
    public async Task<ActionResult<IEnumerable<UserDto>>> GetAllAsync()
    {
        return Ok(await _mediator.Send(new GetAllUsersQuery()));
    }

    [Authorize]
    [HttpGet("/users/getById")]
    public async Task<ActionResult<UserDto?>> GetByIdAsync([FromQuery] int id)
    {
        return Ok(await _mediator.Send(new GetUserByIdQuery { UserId = id }));
    }

    [HttpPost("/users/create")]
    public async Task<IActionResult> CreateUserAsync([FromBody] CreateUserCommand command)
    {
        if (command.Password.Length < 8)
        {
            return BadRequest("Too short password");
        }
        if (!EmailVerifier.IsValidEmail(command.Email))
        {
            return BadRequest("Bad email");
        }
        if (!PhoneNumberVerifier.IsValidPhoneNumber(command.PhoneNumber))
        {
            return BadRequest("Bad phone number");
        }
        if (await _userService.IsDuplicateLoginAsync(command.Login, CancellationToken))
        {
            return BadRequest("Duplicate login");
        }
        return Ok(await _mediator.Send(command));
    }

    [Authorize]
    [HttpPut("/users/update")]
    public async Task<IActionResult> UpdateAsync([FromBody] UpdateUserCommand command)
    {
        if (AuthData.UserId != command.UserId && AuthData.UserRole is not UserRole.Administrator)
        {
            return Unauthorized();
        }

        return Ok(await _mediator.Send(command));
    }

    [Authorize]
    [HttpPut("/users/delete")]
    public async Task<IActionResult> DeleteAsync([FromQuery] int id)
    {
        if (AuthData.UserId != id && AuthData.UserRole is not UserRole.Administrator)
        {
            return Unauthorized();
        }

        return Ok(await _mediator.Send(new DeleteUserCommand { UserId = id }));
    }

    [Authorize(Roles = "Administrator")]
    [HttpPut("/users/setRole")]
    public async Task<IActionResult> SetRoleAsync([FromBody] SetUserRoleCommand command)
    {
        if (command.NewRole is UserRole.Administrator)
        {
            return Unauthorized();
        }

        return Ok(await _mediator.Send(command));
    }

    [Authorize]
    [HttpGet("/users/getId")]
    public async Task<ActionResult<GetIdAndLoginResponse>> GetIdAsync()
    {
        return Ok(await _mediator.Send(new GetIdAndLoginQuery { UserId = AuthData.UserId }));
    }

    [Authorize]
    [HttpGet("/users/getSelfInfo")]
    public Task<ActionResult<UserDto?>> GetSelfInfoAsync() => GetByIdAsync(AuthData.UserId);
}